crates/types/src/predicate/iter/ref_iter.rs

```diff
@@ -0,0 +1,99 @@
+use header::{encode_header, encode_header_unchecked};
+
+use super::*;
+use std::borrow::Cow;
+
+pub struct Iter<'a> {
+    predicate: &'a Predicate,
+    header: Vec<u8>,
+    cursor: Cursor,
+    current: std::slice::Iter<'a, Vec<u8>>,
+}
+
+impl<'a> Iter<'a> {
+    pub fn new(predicate: &'a Predicate) -> Result<Self, core::num::TryFromIntError> {
+        let header = encode_header(predicate)?;
+        Ok(Self::new_inner(predicate, header))
+    }
+
+    pub fn new_unchecked(predicate: &'a Predicate) -> Self {
+        let header = encode_header_unchecked(predicate);
+        Self::new_inner(predicate, header)
+    }
+
+    fn new_inner(predicate: &'a Predicate, header: Vec<u8>) -> Self {
+        Self {
+            predicate,
+            header,
+            cursor: Cursor::Header,
+            current: predicate.state_read.iter(),
+        }
+    }
+    fn state_reads(&mut self) {
+        self.cursor = Cursor::StateReads;
+    }
+
+    fn constraints(&mut self) {
+        self.cursor = Cursor::Constraints;
+        self.current = self.predicate.constraints.iter();
+    }
+
+    fn directive(&mut self) {
+        self.cursor = Cursor::Directive;
+    }
+}
+
+impl<'a> Iterator for Iter<'a> {
+    type Item = Cow<'a, [u8]>;
```

Comment:
A `Cow` is required because the header is owned by the iterator and can't be returned as a reference because it won't have the 'a lifetime.

While not ideal the alternative is to create the header separately and then pass it to the iterator. 
So the best you could do is:
```rust
let header = predicate.header();
let bytes = Iter::new(&predicate, &header).collect();
```
These wouldn't work:
```rust
predicate.iter().collect();
// Or
let iter = Iter::new(&predicate, &predicate.header());
```
So it's kind of the less of two evils